Alternate version of #69: Same as #69 + unit tests + unit test Chart struct

frizzr opened the pull request:
This alternate version of #69 also adds unit tests for the original PR, but I also added the Chart struct to help with, for example, modifying or resetting the content of your test Helm chart(s).

chrira reviewed (changes requested):
Thanks for the tests!
I have some errors running `dagger -m tests/ call all`
```
./main.go:140:145: unknown field Version in struct literal of type dagger.HelmPackagePushOpts                                                                                                  
./main.go:169:145: unknown field AppVersion in struct literal of type dagger.HelmPackagePushOpts   
```
You are not having problems?
Sometime I have testing issues because of caching.

frizzr commented:
My apologies for those issues. 

I am having problems running unit tests in my local environment....in our company we have to use a customized engine image for Dagger due to needing to setup certificates for our company proxy.

RIght now we use Dagger v0.14.0 with our customized base image while your unit tests use v0.16. Do you have any general Dagger advice for handling different environments like this?

frizzr commented:
> You are not having problems?
> Sometime I have testing issues because of caching.

My apologies for those issues.

I am having problems running unit tests in my local environment....in our company we have to use a customized engine image for Dagger due to needing to setup certificates for our company proxy.

RIght now we use Dagger v0.14.0 with our customized base image while your unit tests use v0.16. Do you have any general Dagger advice for handling different environments like this?

chrira commented:
@frizzr no worries!

Yes, the tests are executed in GitHub.
The module tests should run on your fork. The integration test will not because of missing credentials.

Here is one example on your fork: https://github.com/frizzr/dagger-module-helm/actions/runs/13639799085/job/38126851023

Hope this helps you getting forward.

frizzr commented:
Hello @chrira! It has been awhile, but I finally finished a complete rework of the unit tests along with my feature adds. I'm happy to say that it is all safe to execute the tests in parallel and I came up with a way to guarantee each test is _completely independent of each other_ by having a set of `Chart` structs that represents our `testdata` charts in `tests/testdata`.

This struct, when instantiated as part of `NewChartSession`, will automatically rewrite each `testdata` chart's Chart.yaml _inside the container's context_ and automatically suffix a random string to the end of the chart name (without changing the host's content.)

With a different set of chart names per test, it guarantees that we can safely perform tests using `ttl.sh` and all of this implementation is located in `tests/charts.go` hidden away from the actual unit tests. The unit tests of course have access to what the new chart name is within each test.

This was a LOT of work I know, and I hope it isn't too crazy or too drastic of a change. Still, I think it could solve some things now and into the future, but I'll let you be the judge of that.

If the whole thing is "too much", then I could fall back and move my feature adds across to the other PR and leave the unit tests alone.

Thanks for considering!

frizzr commented on helm/main.go:
```diff
@@ -24,6 +24,8 @@ type PushOpts struct {
 	Username   string `yaml:"username"`
 	Password   *dagger.Secret
+	Version    string `yaml:"version"`
+	AppVersion    string `yaml:"appVersion"`
```
Done!

frizzr commented on helm/main.go:
```diff
@@ -102,33 +133,40 @@ func (h *Helm) PackagePush(
-	version = strings.TrimSpace(version)
-
-	name, err := c.WithExec([]string{"sh", "-c", "helm show chart . | yq eval '.name' -"}).Stdout(ctx)
+	name, err := h.queryChartWithYq(ctx, directory, ".name")
```
Yes indeed. Done.

RetGal commented on helm/main.go:
```diff
@@ -268,7 +392,8 @@ func (h *Helm) doesChartExistOnRepo(
 		if exc == "0" {
 			//Chart exists
-			return true, nil
+			return true, fmt.Errorf("Debug")
```
```suggestion
			return true,  nil
```

Should return nil here. 
Returning "Debug" would fail with the original `HelmPackagePushWithExistingChart` test function, but somehow passes with the new test.
Original test: https://github.com/puzzle/dagger-module-helm/blob/main/tests/main.go#L105

frizzr replied:
This has been changed back to `nil`.

chrira commented:
@frizzr what a work, thank you!

Unfortunately it breaks our main return logic of the PackagePush method. See comment:
https://github.com/puzzle/dagger-module-helm/blob/main/helm/main.go#L65

As @RetGal mentioned, the test method for this logic on the main branch (helm-package-push-with-existing-chart) fails with your changes.

Changing `fmt.Errorf("Debug")` back to `nil` and re-adding the test should fix it.

frizzr commented:
@chrira and @RetGal :

I added a good bit more logic (including a use case I needed with Lint, which is we need to run `helm dependency update` to do a full lint) and refactored things to simplify the source code. I also changed the `fmt.Errorf("Debug")` back to `nil`. 

The big question I have is....how did my non-OCI unit test ever work? Wouldn't we need to simulate a non-OCI repository to do the test (given that `ttl.sh` only can do OCI repositories?)

RetGal commented:
@frizzr since you are the only person we know who uses this module along with a non-oci registry and we currently have no means to unit test it, i'd propose to merge it, if it works for your real life use case.

@chrira any objections from your side?

chrira commented on helm/main.go:
```diff
@@ -12,18 +12,23 @@ import (
 	"fmt"
 	"os"
 	"strings"
+    "hash/crc32"
```
fix formatting

chrira commented on helm/main.go:
```diff
@@ -12,18 +12,23 @@ import (
+	Registry   		string `yaml:"registry"`
+	Repository 		string `yaml:"repository"`
+	Oci        		bool   `yaml:"oci"`
+	NonOCISubpath	string `yaml:"nonOciSubpath"`
```
fix alignment (check tabs/spaces)

chrira commented on helm/main.go:
```diff
@@ -101,37 +153,72 @@ func (h *Helm) PackagePush(
+		Registry:		registry,
+		Repository:		repository,
+		Oci:			!useNonOciHelmRepo,
+		NonOCISubpath:	nonOciRepoSubpath,
```
fix alignment (check tabs/spaces)

chrira commented on helm/main.go:
```diff
@@ -101,37 +153,72 @@ func (h *Helm) PackagePush(
-		WithWorkdir("/helm")
-	version, err := c.WithExec([]string{"sh", "-c", "helm show chart . | yq eval '.version' -"}).Stdout(ctx)
+	c := h.createContainer(directory)
+	//c := dag.Container().
```
please remove commented code

chrira commented on helm/main.go:
```diff
@@ -197,7 +293,8 @@ func (h *Helm) Test(
 	c := h.createContainer(directory)
-	out, err := c.WithExec([]string{"sh", "-c", fmt.Sprintf("%s %s", "helm-unittest", strings.Join(args, " "))}).Stdout(ctx)
+	out, err := c.WithExec([]string{"sh", "-c",
+										fmt.Sprintf("%s %s", "helm-unittest", strings.Join(args, " "))}).Stdout(ctx)
```
check formatting

chrira commented on helm/main.go:
```diff
@@ -293,39 +505,53 @@ func (h *Helm) doesChartExistOnRepo(
 }
 
-func (h *Helm) hasMissingDependencies(
+func (h *Helm) queryChartWithYq(
```
nice!

chrira reviewed (changes requested):
@frizzr are you still working on it?
Note when you are finished. Then we will do the final review and merge it.
Thanks for contributing!

chrira commented:
@frizzr 
Please run `gofmt -w helm/main.go` on your go files. We formatted the code on main.
Then you should be able to rebase your branch with main.
We are still struggling with testing, see https://github.com/puzzle/dagger-module-helm/pull/84
Thanks

frizzr commented:
Sorry for the delays. I am getting laid off from my job effective late next
month, but I do want to keep working on this even after my job ends.

The main problem that I am having right now is very specific to our CI
environment. The code as-is works perfectly except in our
'docker-in-docker' environment...we run Actions Runner Controller in a K8s
cluster and I am having a bizarre issue with `helm repo add` not keeping
state...the very next `WithExec` command I run completely loses the dot
config directory for Helm...but in all other environments it seems to work
great.

I'll post more info later.

Thanks!

On Mon, May 12, 2025, 10:38 Christoph Raaflaub ***@***.***>
wrote:

> *chrira* left a comment (puzzle/dagger-module-helm#70)
> <https://github.com/puzzle/dagger-module-helm/pull/70#issuecomment-2872842744>
>
> @frizzr <https://github.com/frizzr>
> Please run gofmt -w helm/main.go on your go files. We formatted the code
> on main.
> Then you should be able to rebase your branch with main.
> We are still struggling with testing, see #84
> <https://github.com/puzzle/dagger-module-helm/pull/84>
> Thanks
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/puzzle/dagger-module-helm/pull/70#issuecomment-2872842744>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGKTEYEQWTHZW7WPKB5ROD26CW7LAVCNFSM6AAAAABYEJGGYSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQNZSHA2DENZUGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
